### PR TITLE
Change UID2-Shared for OptOut Policy Check

### PR DESCRIPTION
**Pull request by zaiweidu**
1. Added a new method `getOldestClientKey()` to RotatingClientKeyProvider.java
2. Added unit tests to RotatingClientKeyProviderTest.java

### CODE REVIEW

File: src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java [hunk 1]
```diff
@@ -93,4 +95,12 @@ public CloudPath getMetadataPath() {
     public IAuthorizable get(String key) {
         return getClientKey(key);
     }
+
+    public ClientKey getOldestClientKey(int siteId) {
+        List<ClientKey> keys = reader.getSnapshot().values().stream()
+                .filter(k -> k.getSiteId() == siteId) // filter by site id
+                .sorted(Comparator.comparing(ClientKey::getCreated)) // sort by key creation timestamp ascending
+                .collect(Collectors.toList());
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java:103]: Would `findFirst()` be more appropriate here?

File: src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java [hunk 2]
```diff
@@ -93,4 +95,12 @@ public CloudPath getMetadataPath() {
     public IAuthorizable get(String key) {
         return getClientKey(key);
     }
+
+    public ClientKey getOldestClientKey(int siteId) {
```
Comment (atarassov-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java:99]: Should this be a method on the snapshot instead? This is not really specific to the rotating provider.

File: src/test/java/com/uid2/shared/auth/RotatingClientKeyProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import com.uid2.shared.store.scope.GlobalScope;
+import io.vertx.core.json.JsonObject;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.*;
+
+public class RotatingClientKeyProviderTest {
+    @Test
+    public void testGetOldestClientKeyBySiteId() throws Exception {
```
Comment (atarassov-ttd) [inline src/test/java/com/uid2/shared/auth/RotatingClientKeyProviderTest.java:15]: A test for when site does not have any keys?

File: src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java (hunk 1 above)
Comment (zaiweidu) [reply src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java]: Replaced `get(0)` with `findFirst()` as suggested.

File: src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java (hunk 2 above)
Comment (zaiweidu) [reply src/main/java/com/uid2/shared/store/reader/RotatingClientKeyProvider.java]: Moved `getOldestClientKey()` to _ClientKeySnapshot_.

File: src/test/java/com/uid2/shared/auth/RotatingClientKeyProviderTest.java (hunk 3 above)
Comment (zaiweidu) [reply src/test/java/com/uid2/shared/auth/RotatingClientKeyProviderTest.java:15]: Test case added as suggested.

**Review by atarassov-ttd — APPROVED** [review]

**Pull request merged**